File: Algorithms/763. Partition Labels/rust/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn partition_labels(s: String) -> Vec<i32> {
        let mut ans = Vec::<i32>::new();
        let mut last = HashMap::<char,_>::new();
        for (i, c) in s.chars().enumerate() {
            last.insert(c, i);
        }
        let mut anchor = 0;
        let mut j = 0;
        for( i, c) in s.chars().enumerate() {
            j = j.max(*last.get(&c).unwrap());
            if i == j {
                ans.push((j - anchor + 1) as i32);
                anchor = j + 1;
            }
        }
        ans
    }
}

File: Algorithms/763. Partition Labels/python/solution.py
class Solution:
    def partitionLabels(self, s: str) -> list[int]:
        last = {c: i for i, c in enumerate(s)}
        j = 0
        anchor = -1
        ans = []
        for i, c in enumerate(s):
            j = max(j, last[c])
            if i == j:
                ans.append(i - anchor)
                anchor = i

        return ans
